Change iterable _userdata to deque

swails opened the pull request:
Some other cosmetic fixes.

Fixes #87 

Signed-off-by: Jason Swails <jason.swails@gmail.com>

swails commented on src/paho/mqtt/client.py:
```diff
@@ -353,10 +353,7 @@ def __ne__(self, other):
-            return self._topic.decode('utf-8')
-        else:
-            return self._topic
+        return self._topic.decode('utf-8')
```
The version check is unnecessary here since `str` in Python 2 accepts `decode` just fine:

```python
Python 2.7.13 |Anaconda 2.5.0 (x86_64)| (default, Dec 20 2016, 23:05:08) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> u'some string'.decode('utf-8')
u'some string'
>>> 'some string'.decode('utf-8')
u'some string'
```

swails commented on src/paho/mqtt/client.py:
```diff
@@ -490,9 +487,14 @@ def __init__(self, client_id="", clean_session=True, userdata=None,
         self._protocol = protocol
-        self._userdata = userdata
+        if isinstance(userdata, collections.Iterable) and not isinstance(userdata, basestring):
+            self._userdata = collections.deque(userdata)
```
docstrings indicate that `userdata` can be any type, but then it's treated inside `_do_publish` as an iterable with a `pop` method.  What should happen here?  Should `userdata` be enforced to be a deque and raise a `TypeError` if it's not an iterable that can be coerced into one?

Nzbuu replied:
The client used in the publish module sets the userdata to be a queue. As you say, in the client module, it should be anything.

swails commented:
Thanks @Nzbuu!  I've made the appropriate changes.

I also changed the imports into relative imports with the appropriate `__future__` feature.  The upside of this is that `paho` can now be copied in its entirety as a subpackage, allowing users to bundle `paho` directly and eliminate it as a dependency *without* colliding with any other installed version of `paho` (or getting silently overwritten if a user then pip-installs paho separately).

There are also two instances where failed `isinstance` checks resulted in a `ValueError` being raised rather than the more sensible `TypeError`.  I changed these to the exception type that makes more sense, but it *does* technically change undocumented API features and potentially introduces some backwards-incompatible changes in a few corner cases.

Comments welcome.

swails commented:
Last change I just made was to replace uses of `print` with logging

swails commented:
Any other issues with this PR?

PierreF commented:
I don't like the logger change part of this PR. It introduce another logger:
* print() (ok this one no longer exists after this PR)
* existing on_log() method, that may using Python logging (en enable_logger is called)
* LOGGER introduced by this PR

I think it's better to use existing logger (_easy_log).

Could you remove the logger change from this PR (it's not releated to the PR, prefer to open multiple PR with one per subject it makes review easier). The other changes (PR related, pep8 fixes, import changes, and exception changes - ideally this would have been 4 PRs) looks good to me.

swails commented:
I made the logging change, now I'll split this into separate PRs.
